### PR TITLE
[SPARK-9149][ML][Examples] Add an example of spark.ml KMeans

### PR DESCRIPTION
**Pull request by yu-iskw**
[SPARK-9149] Add an example of spark.ml KMeans - ASF JIRA https://issues.apache.org/jira/browse/SPARK-9149

@jkbradley Should we support other data formats, such as TSV or CSV. I have implemented these examples which support only space separated file which is same as the example for `spark.mllib`'s `KMeans`.

### CODE REVIEW

**Comment by yu-iskw** [comment]
Jenkins, retest this please.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.examples.ml;
+
+import java.util.regex.Pattern;
+
+import org.apache.commons.cli.*;
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java]: I think we should not introduce a dependency on Commons CLI just for this. It's ancient. Actually, I see one other use of this in an example, which shouldn't be there as this is an undeclared dependency. It's not crazy to fix that here; it could be a separate PR though.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+import org.apache.spark.sql.catalyst.expressions.GenericRow;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+
+/**
+ * An example demonstrating a k-means clustering.
+ * Run with
+ * {{{
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java]: Unless I'm really out of date, this doesn't work in Javadoc. `{@code ...}` works. Though that is not used in Spark consistently.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    Integer maxIter = 20;
+    Integer runs = 1;
+    Double epsilon = 1E-4;
+    Long seed = 1L;
+    String initMode = KMeans.K_MEANS_PARALLEL();
+    Integer initSteps = 5;
+  }
+
+  private static class ParsePoint implements Function<String, Row> {
+    Pattern separater = Pattern.compile(" ");
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java]: Nit: "separator". This should be private static if possible.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java (hunk 2 above)
Comment (yu-iskw) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java]: I confused it with Scaladocs. I'll fix it. Thanks.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java (hunk 3 above)
Comment (yu-iskw) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java]: Alright.

**Comment by yu-iskw** [comment]
@srowen I made examples in Scala and Java more simple. Could you review it?

File: examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+ * An example demonstrating a k-means clustering.
+ * Run with
+ * <pre>
+ * bin/run-example ml.JavaSimpleParamsExample <file> <k>
+ * </pre>
+ */
+public class JavaKMeansExample {
+
+  private static class ParsePoint implements Function<String, Row> {
+    private static Pattern separater = Pattern.compile(" ");
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java]: Nit: The spelling should still be `separator` and it can be `final`.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Override
+    public Row call(String line) {
+      String[] tok = separater.split(line);
+      double[] point = new double[tok.length];
+      for (int i = 0; i < tok.length; ++i) {
+        point[i] = Double.parseDouble(tok[i]);
+      }
+      Vector[] points = {Vectors.dense(point)};
+      Row row = new GenericRow(points);
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java]: While we're here, this can be returned directly, skipping a local var

File: examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    // Parses the arguments
+    SparkConf conf = new SparkConf().setAppName("JavaKMeansExample");
+    JavaSparkContext jsc = new JavaSparkContext(conf);
+    SQLContext sqlContext = new SQLContext(jsc);
+
+    // Loads data
+    JavaRDD<Row> points = jsc.textFile(inputFile).map(new ParsePoint());
+    StructField[] fields = new StructField[1];
+    fields[0] = new StructField("features", new VectorUDT(), false, Metadata.empty());
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java]: You can use the same `Foo[] = { ... };` declaration as above here?

File: examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    StructType schema = new StructType(fields);
+    DataFrame dataset = sqlContext.createDataFrame(points, schema);
+
+    // Trains a k-means model
+    org.apache.spark.ml.clustering.KMeans kmeans = new org.apache.spark.ml.clustering.KMeans()
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java]: Can this be an import or am I missing why it has to be fully qualified here?

File: examples/src/main/scala/org/apache/spark/examples/ml/KMeansExample.scala [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+    // Creates a Spark context and a SQL context
+    val conf = new SparkConf().setAppName(s"${this.getClass.getSimpleName}")
+    val sc = new SparkContext(conf)
+    val sqlContext = new SQLContext(sc)
+    import org.apache.spark.mllib.linalg.VectorUDT
+
+    // Loads data
+    val rowRDD = sc.textFile(input).filter(l => l != "")
+      .map(_.split(" ").map(v => java.lang.Double.parseDouble(v)))
```
Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/ml/KMeansExample.scala]: `_.toDouble` instead of using `java.lang.Double`?

Comment (srowen) [inline examples/src/main/scala/org/apache/spark/examples/ml/KMeansExample.scala]: I think the filter condition can be tightened to `_.nonEmpty`? and likewise below can you `map(Vectors.dense)`? I forget whether that syntax will work.

**Comment by srowen** [comment]
LGTM pending tests; wouldn't hurt to have @jkbradley look.

**Comment by yu-iskw** [comment]
@srowen Thank you for reviewing it!
@jkbradley Could you take a glance at this?

File: examples/src/main/scala/org/apache/spark/examples/ml/KMeansExample.scala [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+ */
+object KMeansExample {
+
+  final val FEATURES_COL = "features"
+
+  def main(args: Array[String]): Unit = {
+    if (args.length != 2) {
+      // scalastyle:off println
+      System.err.println("Usage: ml.KMeansExample <file> <k>")
+      // scalastyle:of println
```
Comment (techaddict) [inline examples/src/main/scala/org/apache/spark/examples/ml/KMeansExample.scala]: I think you meant to write `scalastyle:on println`.

File: examples/src/main/scala/org/apache/spark/examples/ml/KMeansExample.scala [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+      System.exit(1)
+    }
+    val input = args(0)
+    val k = args(1).toInt
+
+    // Creates a Spark context and a SQL context
+    val conf = new SparkConf().setAppName(s"${this.getClass.getSimpleName}")
+    val sc = new SparkContext(conf)
+    val sqlContext = new SQLContext(sc)
+    import org.apache.spark.mllib.linalg.VectorUDT
```
Comment (techaddict) [inline examples/src/main/scala/org/apache/spark/examples/ml/KMeansExample.scala]: why not import this at the beginning ?

**Comment by yu-iskw** [comment]
@techaddict thank you for your comments.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+ * An example demonstrating a k-means clustering.
+ * Run with
+ * <pre>
+ * bin/run-example ml.JavaSimpleParamsExample <file> <k>
+ * </pre>
+ */
+public class JavaKMeansExample {
+
+  private static class ParsePoint implements Function<String, Row> {
+    final private static Pattern separator = Pattern.compile(" ");
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaKMeansExample.java]: This is picking nits, and something we can fix on merge, but the normal order of modifiers is `private static final ...`

**Comment by srowen** [comment]
I think this is pretty fine, minus one thing I can fix on merge. Any more comments?

**Comment by jkbradley** [comment]
@yu-iskw Thanks for adding this, and others for reviewing!  It looks good.  My only comment is that it might be good to use the built-in MLlib methods like MLUtils.loadVectors to load data, rather than having new parsing methods in examples. Not a big deal though.

**Comment by yu-iskw** [comment]
@srowen thank you for merging it!

@jkbradley thank you for your feedback! I agree with that it would be better to use `MLUtils.loadVectors`. However,  it doesn't support space separated format. So I was wondering if I should keep consistency with the input data format of the `spark.mllib KMeans` example or should create new one regardless of it. Finally, I thought it would be better to keep consistency with the `spark.mllib KMeans` example.

**Comment by jkbradley** [comment]
Ohh, I see.  That's fine.  Thanks!
